R/dependecies.R

```diff
@@ -0,0 +1,510 @@
+        package = package
+        , dependencies = single_package_dependencies)
+    } else {
+      single_package_dependencies <-
+        previously_parsed_deps[[previously_parsed_loc]]$dependencies
+    }
+    current_dependencies <- combine_dependencies(
+      single_package_dependencies
+      , current_dependencies
+      , package$name)
+  }
+  if (identical(master_list, current_dependencies)) return(master_list)
+  get_dependencies_for_list(current_dependencies, lock, previously_parsed_deps)
+}
+
+#' Where, if anywhere, have I stored this packages/version combinations'
+#' previously parsed dependencies?
+which_previously_parsed <- function(package, previously_parsed_deps) {
+  if (length(previously_parsed_deps) == 0) return(0L)
+  previously_parsed_names <- vapply(previously_parsed_deps
+    , function(p) p[["package"]]$name, character(1))
+  if (package$name %in% previously_parsed_names) {
+    sel <- which(previously_parsed_names == package$name)
+    subsel <- vapply(previously_parsed_deps[sel]
+      , function(p) identical(p[["package"]]$version, package$version)
+      , logical(1))
+    if (any(subsel)) {
+      return(sel[subsel])
+    }
+  }
+  0L
+}
+
+#' Attach the latest available lockbox version to a package
+add_latest_version_in_lockbox <- function(package) {
+  package$latest_version_in_lockbox <- max_package_version(
+    list.files(file.path(lockbox_library(), package)))
+  package
+}
+
+#' Take a vector of versions and find the max without coercing to package_version
+max_package_version <- function(versions) {
+  if (length(versions) == 0) return(NULL)
+  formatted_versions <- package_version(as.character(versions))
+  versions[which(formatted_versions == max(formatted_versions))[1]]
+}
+
+#' Check a dependency list for inclusion in the lockfile and replace the package
+#' with the locked version if it does appear there.  Also throw
+#' an error if we require a dependency version greater than that specified by
+#' the lockfile.
+replace_with_lock <- function(package, lock) {
+  lock_names <- vapply(lock, function(l) l$name, character(1))
+  if (package$name %in% lock_names) {
+    locked_package <- lock[[which(lock_names == package$name)[1]]]
+    if (!is.na(package$version) && package_version(as.character(package$version)) >
+      package_version(as.character(locked_package$version))) {
+        stop(paste0("Dependency: \'", package$name, ", Version: ", package$version
+          , " is required, but lockbox is locked at version: "
+          , as.character(locked_package$version)))
+    }
+    package <- locked_package
+  } else {
+    package$is_dependency_package <- TRUE
+  }
+  if (!"remote" %in% names(package) || is.na(package$remote)) {
+    package$remote <- "CRAN"
+  }
+  package <- as.locked_package(package)
+  if (package$is_dependency_package && !package$name %in%
+    as.character(installed.packages(priority = "base")[,1])) {
```

Comment:
There are packages that are integrated into base now.  Trying to find them is a waste of time and a high potential for errors